### PR TITLE
Improve card header UI

### PR DESCRIPTION
**Pull request by jart**
This change improves the aesthetics of figure captions. It makes the information seem less crowded, while also freeing up vertical space. The improvement looks the best on the image dashboard.

Before:

image

After:

image

### CODE REVIEW

File: tensorboard/components/tf_card_heading/tf-card-heading-style.html [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+<!--
+  Shared stylesheet for figure captions.
+-->
+<dom-module id="tf-card-heading-style">
+  <template>
+    <style>
+
+      /** Horizontal line of labels. */
+      .row {
```
Comment (chihuahua) [inline tensorboard/components/tf_card_heading/tf-card-heading-style.html]: This CSS is included within various modules, and it affects all elements within the modules. Could we prefix these declarations with `tf-card-heading`? ie,

```css
tf-card-heading .row {
   ...
}
```
etc. 

to prevent the declarations from affecting unrelated elements (say also with the class `row` or `label`).

File: tensorboard/components/tf_card_heading/tf-card-heading-style.html [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      .label {
+        float: left;
+        margin-top: 4px;
+        /* TODO(@jart): Use proper width so overflow works. */
+        text-overflow: ellipsis;
+        overflow: hidden;
+      }
+
+      /** Makes label show on the right. */
+      .right {
```
Comment (chihuahua) [inline tensorboard/components/tf_card_heading/tf-card-heading-style.html]: It's generally recommended to name classes after their semantic purpose instead of how they will be rendered (ie, positioned on the right side).

That's just a recommendation though, and this case seems rather special because there's no other way to describe something on the right side, so this might be a time to go against that. Maybe `right-side-label`?

File: tensorboard/components/tf_card_heading/tf-card-heading.html [hunk 3]
```diff
@@ -62,25 +82,25 @@
...
-      }
-      .label:first-child {
+      .name {
         font-size: 14px;
       }
+      .run {
+        font-size: 11px;
+        width: auto;
+        border-radius: 3px;
+        text-shadow: 0 -1px 0 rgba(0,0,0,0.25);
```
Comment (chihuahua) [inline tensorboard/components/tf_card_heading/tf-card-heading.html]: FYI, I personally avoid text-shadow due to performance issues.
https://stackoverflow.com/questions/32948596/is-csss-efficiency-with-text-shadow-as-bad-as-box-shadow

but this is a small text shadow, and our frontend is not bottlenecked by browser paints, so I think we're good.

File: tensorboard/components/tf_card_heading/util.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * Formats timestamp for the card header.
+ * @param {Date} date
+ * @return {string}
+ */
+export function formatDate(date) {
+  if (!date) {
+    return '';
+  }
+  // Turn things like "GMT-0700 (PDT)" into just "PDT".
+  return date.toString().replace(/GMT-\d+ \(([^)]+)\)/, '$1');
```
Comment (chihuahua) [inline tensorboard/components/tf_card_heading/util.js]: Nice date formatting.

File: tensorboard/components/tf_card_heading/util.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+
+/**
+ * Returns CSS color that will contrast against background.
+ * @param {?string} background
+ * @return {string}
+ */
+export function pickTextColor(background) {
```
Comment (chihuahua) [inline tensorboard/components/tf_card_heading/util.js]: This implementation of finding a contrasting color is really cool!

File: tensorboard/components/tf_card_heading/util.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    return '';
+  }
+  // Turn things like "GMT-0700 (PDT)" into just "PDT".
+  return date.toString().replace(/GMT-\d+ \(([^)]+)\)/, '$1');
+}
+
+
+/**
+ * Returns CSS color that will contrast against background.
+ * @param {?string} background
```
Comment (chihuahua) [inline tensorboard/components/tf_card_heading/util.js]: Oh, and maybe document that this should be a hex color. CSS colors as you probably know can come in many forms (RGB(A), HSL, colloquial name, etc).

File: tensorboard/components/tf_card_heading/tf-card-heading-style.html (hunk 1 above)
Comment (jart) [reply tensorboard/components/tf_card_heading/tf-card-heading-style.html]: Done.

File: tensorboard/components/tf_card_heading/tf-card-heading-style.html (hunk 2 above)
Comment (jart) [reply tensorboard/components/tf_card_heading/tf-card-heading-style.html]: I kind of like right because when you write the element it says `class="label right"`.

File: tensorboard/components/tf_card_heading/tf-card-heading.html (hunk 3 above)
Comment (jart) [reply tensorboard/components/tf_card_heading/tf-card-heading.html]: Noted.

File: tensorboard/components/tf_card_heading/util.js (hunk 4 above)
Comment (jart) [reply tensorboard/components/tf_card_heading/util.js]: Danke

File: tensorboard/components/tf_card_heading/util.js (hunk 6 above)
Comment (jart) [reply tensorboard/components/tf_card_heading/util.js]: Done.

File: tensorboard/components/tf_card_heading/util.js (hunk 5 above)
Comment (jart) [reply tensorboard/components/tf_card_heading/util.js]: I agree. So glad the w3c documented this.

**Comment by wchargin** [comment]
Are the following phenomena intentional?
  * Top: the tag name is not left-aligned but not centered, either, though it has `float: left` (clearfix needed?)
  * Middle: long-ish run name is the only right-aligned thing, yielding a weird back-and-forth flow.
  * Bottom: all components run together on text dashboard (because although they are `display: block` they are also `float: none`.)

I've found that everything is much simpler when we avoid the `float` property. For example, the little "info" icon for the summary descriptions could have been `float: right`, but letting it have a separate table cell works very well. (Exception: actual floating figures in article-like text..)

Screenshot with the first two mentioned phenomena.

Screenshot with the last mentioned phenomenon.


I really like the colors, though.

**Comment by chihuahua** [comment]
Ah! I deem 1 and 3 to be bugs.

For 1, we should introduce a container with `clear: both` after the 1st row. For 3, we should maybe add some left margin on the right label so that we forbid text from running into each other like that (and instead put them in different rows).

For 2, I think the intention was to show the right label on a new line if it is too long.

**Comment by wchargin** [comment]
Okay: that behavior for (2) is all right with me (certainly the least disagreeable). I'd probably prefer it to be aligned left if it has to wrap, but I appreciate that that's probably not worth the CSS hacks that would be required to implement it.

Perhaps the `class="row"`s are supposed to be `class="heading-row"`, which would give them `clear: both`? (But also a negative margin. Not sure what's desired here.)

I don't understand your suggestion about adding a `margin-left` to the component with "tag: higher_order_tensors/multiplication_table". This is one potential solution if we want to keep it on the same line as the display name—is this really what we want?—but that margin would persist even when the tag wraps to a new line, and we certainly don't want it there. Why don't we just force each component to be in its own row? I suspect that the fix from (1) may also work here.

**Comment by jart** [comment]
Yes I agree it would be nice if it could align left without a margin when it wraps, although I wasn't sure how to do that off the top of my head. Also yes, class=row should be heading-row. That's a refactoring bug.

**Comment by wchargin** [comment]
[How about this?](https://react.jsbin.com/naqumoj/1/edit?js,output)

Result

TL;DR: `flex-grow: 1` on the display name, `flex-grow: 0` on the "run name pill".

**Comment by wchargin** [comment]
[How about this?](https://react.jsbin.com/naqumoj/1/edit?js,output)

Result

TL;DR: `flex-grow: 1` on the display name, `flex-grow: 0` on the "run name pill". This also does not require any `float`s.

**Comment by chihuahua** [comment]
I like it! And then, we have 2 flex containers - 1 for each row.

**Comment by chihuahua** [comment]
FYI, somebody should create a PR and make this happen.
